main.cpp

```diff
@@ -1,15 +1,155 @@
 #include <getopt.h>
 #include <iostream>
+#include <string>
 #include <cstdlib>
+// #include <cstdio>
+// #include <stdexcept>
 
 #include "helpers.h"
 
 using namespace std;
 
 // main file
-int main(int argc, char **argv)
-{
-    // Code to be written below this comment
+int main(int argc, char **argv) {
+    try {
+        // Define allowable filters
+        const char *filtersOptions {"begrs"};
 
+        // Get filter flag
+        char filter {static_cast<char>(getopt(argc, argv, filtersOptions))};
+
+        // check validity
+        if (filter == '?') {
+            throw string {"Invalid filter."};
+        }
+
+        // Ensure only one filter
+        if (getopt(argc, argv, filtersOptions) != -1) {
+            throw string {"Only one filter allowed."};
+        }
+
+        // Ensure proper usage
+        if (argc != optind + 2) {
+            throw string {"Usage: filter [flag] infile outfile"};
+        }
+
+        // Open input file 
+        FILE *inFile {fopen(argv[optind], "r")};
+        // check readability
+        if (inFile == nullptr) {
+            throw string {"Could not read : " + static_cast<string>(argv[optind])};
+        }
+
+        // open output file
+        FILE *outFile {fopen(argv[optind + 1], "w")};
+        // check writability
+        if (outFile == nullptr) {
+            throw string {"Could not create : " + static_cast<string>(argv[optind + 1])};
+        }
+
+        // Read inFile's BITMAPFILEHEADER
+        BITMAPFILEHEADER bf;
+        fread(&bf, sizeof(BITMAPFILEHEADER), 1, inFile);
+
+        // Read inFile's BITMAPINFOHEADER
+        BITMAPINFOHEADER bi;
+        fread(&bi, sizeof(BITMAPINFOHEADER), 1, inFile);
+
+        // Ensure infile is (likely) a 24-bit uncompressed BMP 4.0
+        if (bf.bfType != 0x4d42 || bf.bfoffBits != 54 || bi.biSize != 40 || bi.biBitCount != 24 || bi.biCompression != 0) {
+            fclose(outFile);
+            fclose(inFile);
+            
+            throw string {"Unsupported file format."};
+        }
+
+        int height {abs(bi.biHeight)};
+        int width {abs(bi.biWidth)};
+
+        // Allocate memory for image
+        RGBTRIPLE *image {new RGBTRIPLE[height * width]};
```

Comment:
```c
RGBTRIPLE(*image)[width] = calloc(height, width * sizeof(RGBTRIPLE));
```
Here `pointer to an array` is used or in short `array pointer`. Is this similar to what we are expecting from the new code?